backend/src/util/http_reader.rs

```diff
@@ -0,0 +1,298 @@
+                this.http_client.clone(),
+                this.http_url.clone(),
+                *this.cursor_pos,
+                buf.remaining(),
+                *this.total_bytes,
+            )
+            .boxed()
+        };
+
+        let res_poll = fut.as_mut().poll(cx);
+        println!("polling");
+        println!("buf remaining: {}", buf.remaining());
+        match res_poll {
+            Poll::Ready(result) => match result {
+                Ok(data_chunk) => {
+                    println!("data chunk: len: {}", data_chunk.len());
+                    println!("buf filled len: {}", buf.filled().len());
+                    if data_chunk.len() <= buf.remaining() {
+                        buf.put_slice(&data_chunk);
+                        *this.cursor_pos += data_chunk.len();
+
+                        Poll::Ready(Ok(()))
+                    } else {
+                        println!("data chunk: len: {}", data_chunk.len());
+                        println!("buf filled len: {}", buf.filled().len());
+                        buf.put_slice(&data_chunk);
+
+                        Poll::Ready(Ok(()))
+                    }
+                }
+                Err(err) => Poll::Ready(Err(StdIOError::new(
+                    std::io::ErrorKind::Interrupted,
+                    Box::<dyn std::error::Error + Send + Sync>::from(err.source),
+                ))),
+            },
+            Poll::Pending => {
+                println!("Pending...");
+                *this.read_in_progress = Some(fut);
+
+                Poll::Pending
+            }
+        }
+    }
+}
+
+impl AsyncSeek for HttpReader {
+    fn start_seek(self: Pin<&mut Self>, position: std::io::SeekFrom) -> std::io::Result<()> {
+        let this = self.project();
+
+        match position {
+            std::io::SeekFrom::Start(offset) => {
+                let pos_res = usize::try_from(offset);
+
+                match pos_res {
+                    Ok(pos) => {
+                        if pos > *this.total_bytes {
+                            StdIOError::new(
+                                std::io::ErrorKind::InvalidInput,
+                                format!(
+                                    "position: {} cannot be greater than {} bytes",
+                                    pos, *this.total_bytes
+                                ),
+                            );
+                        }
+                        *this.cursor_pos = pos;
+                    }
+                    Err(err) => return Err(StdIOError::new(std::io::ErrorKind::InvalidInput, err)),
+                }
+                Ok(())
+            }
+            std::io::SeekFrom::Current(offset) => {
+                // We explicitly check if we read before byte 0.
+                let new_pos = i64::try_from(*this.cursor_pos)
+                    .map_err(|err| StdIOError::new(std::io::ErrorKind::InvalidInput, err))?
+                    + offset;
```

Comment:
this will only fail if this.cursor_pos >= 2^63
I'm not saying this isn't worth handling, but I'm also not saying it is

Reply:
Yeah, but since this is an OS basically, my logic was that this code could be ported to other archs or to 32 bits where 2^63 might not be the case?

What do you think?

Reply:
in that case it would be actually impossible to error, because the target is `i64`, not the source.